File: tests/ten_runtime/integration/go/call_api_after_deinited_go/call_api_after_deinited_go_app/ten_packages/extension/default_extension_go/extension.go
//
// Copyright © 2025 Agora
// This file is part of TEN Framework, an open source project.
// Licensed under the Apache License, Version 2.0, with certain conditions.
// Refer to the "LICENSE" file in the root directory for more information.
//

package default_extension_go

import (
	ten "ten_framework/ten_runtime"
)

type serverExtension struct {
	ten.DefaultExtension

	registerCount int
	isStopping    bool
}

func (ext *serverExtension) OnConfigure(tenEnv ten.TenEnv) {
	tenEnv.Log(ten.LogLevelDebug, "OnConfigure")

	ext.registerCount = 0
	ext.isStopping = false
	tenEnv.OnConfigureDone()
}

func (ext *serverExtension) OnStart(tenEnv ten.TenEnv) {
	tenEnv.Log(ten.LogLevelDebug, "OnStart")

	cmd, _ := ten.NewCmd("greeting")
	tenEnv.SendCmd(cmd, nil)

	tenEnv.OnStartDone()
}

func (ext *serverExtension) OnCmd(tenEnv ten.TenEnv, cmd ten.Cmd) {
	name, _ := cmd.GetName()
	if name == "register" {
		ext.registerCount++

		cmdResult, _ := ten.NewCmdResult(ten.StatusCodeOk, cmd)
		tenEnv.ReturnResult(cmdResult, nil)
	} else if name == "unregister" {
		ext.registerCount--

		cmdResult, _ := ten.NewCmdResult(ten.StatusCodeOk, cmd)
		tenEnv.ReturnResult(cmdResult, nil)

		if ext.registerCount == 0 && ext.isStopping {
			tenEnv.OnStopDone()
		}
	} else if name == "test" {
		cmdResult, _ := ten.NewCmdResult(ten.StatusCodeOk, cmd)
		cmdResult.SetPropertyString("detail", "ok")
		tenEnv.ReturnResult(cmdResult, nil)
	} else if name == "hang" {
		// Do nothing
	} else {
		panic("unknown cmd name: " + name)
	}
}

func (ext *serverExtension) OnStop(tenEnv ten.TenEnv) {
	ext.isStopping = true

	if ext.registerCount == 0 {
		tenEnv.OnStopDone()
		return
	}

	tenEnv.Log(ten.LogLevelInfo, "server extension is stopped")
}

type clientExtension struct {
	ten.DefaultExtension
}

func (ext *clientExtension) OnStart(tenEnv ten.TenEnv) {
	tenEnv.Log(ten.LogLevelDebug, "OnStart")

	cmd, _ := ten.NewCmd("register")
	tenEnv.SendCmd(
		cmd,
		func(tenEnv ten.TenEnv, cmdResult ten.CmdResult, err error) {
			if err != nil {
				panic("Failed to send cmd: " + err.Error())
			}

			statusCode, _ := cmdResult.GetStatusCode()
			if statusCode != ten.StatusCodeOk {
				panic("Failed to register")
			}

			tenEnv.OnStartDone()
		},
	)
}

func (ext *clientExtension) OnCmd(tenEnv ten.TenEnv, cmd ten.Cmd) {
	cmdName, _ := cmd.GetName()

	if cmdName == "test" {
		// bypass the cmd to the next extension
		tenEnv.SendCmd(
			cmd,
			func(tenEnv ten.TenEnv, cmdResult ten.CmdResult, err error) {
				if err != nil {
					panic("Failed to send cmd: " + err.Error())
				}

				tenEnv.ReturnResult(cmdResult, nil)
			},
		)
	} else if cmdName == "greeting" {
		cmd, _ := ten.NewCmd("hang")
		tenEnv.SendCmd(cmd, func(te ten.TenEnv, cr ten.CmdResult, err error) {
			statusCode, _ := cr.GetStatusCode()
			if statusCode != ten.StatusCodeError {
				panic("Should receive error status code")
			}

			err = tenEnv.SetProperty("test", true)
			if err == nil {
				panic("Should receive error when setting property")
			}
		})
	} else {
		panic("unknown cmd name: " + cmdName)
	}
}

func (ext *clientExtension) OnDeinit(tenEnv ten.TenEnv) {
	tenEnv.Log(ten.LogLevelDebug, "OnDeinit")

	cmd, _ := ten.NewCmd("unregister")
	tenEnv.SendCmd(
		cmd,
		func(tenEnv ten.TenEnv, cmdResult ten.CmdResult, err error) {
			if err != nil {
				panic("Failed to send cmd: " + err.Error())
			}

			statusCode, _ := cmdResult.GetStatusCode()
			if statusCode != ten.StatusCodeOk {
				panic("Failed to unregister")
			}

			tenEnv.OnDeinitDone()

			err = tenEnv.SetProperty("test", true)
			if err == nil {
				panic("Should receive error when setting property")
			}

			err = tenEnv.Log(ten.LogLevelInfo, "test")
			if err == nil {
				panic("Should receive error when logging")
			}
		},
	)
}

func newAExtension(name string) ten.Extension {
	if name == "server" {
		return &serverExtension{}
	} else if name == "client" {
		return &clientExtension{}
	}

	return nil
}

func init() {
	// Register addon.
	err := ten.RegisterAddonAsExtension(
		"default_extension_go",
		ten.NewDefaultExtensionAddon(newAExtension),
	)
	if err != nil {
		panic("Failed to register addon.")
	}
}

File: tests/ten_runtime/integration/python/multi_process_python/multi_process_python_app/ten_packages/extension/default_extension_python/extension.py
#
# Copyright © 2025 Agora
# This file is part of TEN Framework, an open source project.
# Licensed under the Apache License, Version 2.0, with certain conditions.
# Refer to the "LICENSE" file in the root directory for more information.
#
from multiprocessing import Process
import os
from typing import Optional
from ten_runtime import (
    Extension,
    TenEnv,
    Cmd,
    StatusCode,
    CmdResult,
    TenError,
    LogLevel,
)


class DefaultExtension(Extension):
    def on_configure(self, ten_env: TenEnv) -> None:
        ten_env.log(LogLevel.INFO, "on_init")

        ten_env.init_property_from_json('{"testKey": "testValue"}')
        ten_env.on_configure_done()

    def on_start(self, ten_env: TenEnv) -> None:
        ten_env.log(LogLevel.DEBUG, "on_start")

        ten_env.set_property_from_json("testKey2", '"testValue2"')
        testValue, _ = ten_env.get_property_to_json("testKey")
        testValue2, _ = ten_env.get_property_to_json("testKey2")
        ten_env.log(
            LogLevel.INFO, f"testValue: {testValue}, testValue2: {testValue2}"
        )

        ten_env.on_start_done()

    def on_stop(self, ten_env: TenEnv) -> None:
        ten_env.log(LogLevel.INFO, "on_stop")
        ten_env.on_stop_done()

    def on_deinit(self, ten_env: TenEnv) -> None:
        ten_env.log(LogLevel.INFO, "on_deinit")
        ten_env.on_deinit_done()

    def check_hello(
        self,
        ten_env: TenEnv,
        result: Optional[CmdResult],
        error: Optional[TenError],
        receivedCmd: Cmd,
    ):
        if error is not None:
            assert False, error

        assert result is not None

        statusCode = result.get_status_code()
        detail, _ = result.get_property_string("detail")
        ten_env.log(
            LogLevel.INFO,
            "check_hello: status:" + str(statusCode) + " detail:" + detail,
        )

        respCmd = CmdResult.create(StatusCode.OK, receivedCmd)
        respCmd.set_property_string("detail", detail + " nbnb")
        ten_env.log(LogLevel.INFO, "create respCmd")

        ten_env.return_result(respCmd)

    def on_cmd(self, ten_env: TenEnv, cmd: Cmd) -> None:
        ten_env.log(LogLevel.INFO, "on_cmd")

        cmd_json, _ = cmd.get_property_to_json()
        ten_env.log(LogLevel.INFO, "on_cmd json: " + cmd_json)

        new_cmd = Cmd.create("hello")
        new_cmd.set_property_from_json("test", '"testValue2"')
        test_value, _ = new_cmd.get_property_to_json("test")
        ten_env.log(LogLevel.INFO, "on_cmd test_value: " + test_value)

        p = Process(target=f, args=("bob",))
        p.start()
        p.join()

        ten_env.send_cmd(
            new_cmd,
            lambda ten_env, result, error: self.check_hello(
                ten_env, result, error, cmd
            ),
        )


def info(title):
    print(title)
    print("module name:", __name__)
    print("parent process:", os.getppid())
    print("process id:", os.getpid())


def f(name):
    info("function f")
    print("hello", name)
